### ISSUE TITLE
Nodes classification in multiple graphs 

### ISSUE THREAD

**Issue by lvaleriu** [CrawlScript/tf_geometric#3]
Hello. Nice work, really nice work. I always wondered when TF will have an equivalent of pytorch_geometric lib. It happened and thank you for your effort.

I'm trying to classify graphs nodes in N categories, but I cannot manage to change/combine the demo_mean_pool.py (where you classify the entire graph) and demo_gcn (where you classify nodes in a SINGLE graph) to make it work.

More context:

- input: graphs here are text documents, their nodes are elements of the text document and the edges are built using the spatial location of the elements in the document.  

- output: i'd like to classify each node into a category

Do you have a hint for that?

Thank you!

**Comment by hujunxianligong** [comment]
Thank you for your appreciation!

In the demo, we manually set the `graph.y` to the graph's category with `y=graph_dict["graph_label"]`
```python
def construct_graph(graph_dict):
    return tfg.Graph(
        x=convert_node_labels_to_one_hot(graph_dict["node_labels"]),
        edge_index=graph_dict["edge_index"],
        y=graph_dict["graph_label"]  # graph_dict["graph_label"] is a list with one int element
    )
```

Note that tfg framework does not constrain  `graph.y` to be the label of the graph or something else. `graph.y` can be anything that can be concatenated along the first axis. Thus you can set `graph.y` to the node labels of the graph, for example:
```python
def construct_graph(graph_dict):
    return tfg.Graph(
        x=convert_node_labels_to_one_hot(graph_dict["node_labels"]),
        edge_index=graph_dict["edge_index"],
        y=graph_dict["node_labels"]
    )
```

In this case, you do not need any pooling operation and you can obtain the loss based on the output of the network and  `BatchGraph.y`.

However, I think you should not use graph_dict["node_labels"] as `graph.x` in this case. For featureless nodes, you should consider use dummy node features (such as [1]) or the degree of nodes as node features.

I wish it could help you.

**Comment by lvaleriu** [comment]
Yep, I figured it meanwhile. I've removed the pooling and dense operations and I'm taking the output of the encoder directly (I'll add a BiLSTM after the encoder in a second time). 

    ```
    h = gcn0([batch_graph.x, batch_graph.edge_index, batch_graph.edge_weight], cache=batch_graph.cache)
    h = dropout(h, training=training)
    h = gcn1([h, batch_graph.edge_index, batch_graph.edge_weight], cache=batch_graph.cache)
    return h

I'll be back with questions if needed. Can I leave this issue open for 1, 2 weeks so that I can add new remarks? I'll close it after.

FYI: I'm trying to implement a "BiLSTM NER" that use GCN for exploiting non-local and non-sequentiel context like in this very nice project/paper: 

- https://github.com/thomas0809/GraphIE
- https://arxiv.org/pdf/1810.13083.pdf

I wanted to do that for a long time but didnt find a nice gnn framework before.

Are you planning to add such examples/use cases? It would be great since GCNs used in "sequence taggers"/information extractors are becoming a trend.

**Comment by lvaleriu** [comment]
They actually use a BILSTM + GCN + BILSTM. But I'm not sure how that could be done with tf_geometric. I mean, transform the 1st BILSTM output to a graph input to GCN?

**Comment by hujunxianligong** [comment]
OK, I'll leave this open for 1-2 weeks.

First, most tfg API does not take the Graph object as input. Most operations or layers requires you to pass into the node features (x) and edge information (edge_index). Specifically, the node features can be organized by a matrix.

Note that, the Graph object and BatchGraph object are not necessary for building inputs. They are just for easy organization of data. And you can use it to organize not only the input data, but also the intermediate data.

The Graph object can take A Tensor as `graph.x` and the gradients will be passed to the input. Similarly, building a BatchGraph object with Graphs will not break the gradients. 

Overall, you can use tfg operations, layers and Graph/BatchGraph construction functions as common TensorFlow operations/layers, which are differentiable.

**Comment by hujunxianligong** [comment]
I'll be very happy if someone could contribute such examples.
